pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java

```diff
@@ -0,0 +1,594 @@
+      dataOutputStream.write(dictionaryMapBytes);
+    }
+    // Write data schema bytes.
+    if (dataSchemaBytes != null) {
+      dataOutputStream.write(dataSchemaBytes);
+    }
+    // Write fixed size data bytes.
+    if (_fixedSizeDataBytes != null) {
+      dataOutputStream.write(_fixedSizeDataBytes);
+    }
+    // Write variable size data bytes.
+    if (_variableSizeDataBytes != null) {
+      dataOutputStream.write(_variableSizeDataBytes);
+    }
+
+    // Update the value of "executionThreadCpuTimeNs" to account data table serialization time.
+    long responseSerializationCpuTimeNs = threadTimer.stopAndGetThreadTimeNs();
+    long executionThreadCpuTimeNs =
+        Long.parseLong(getMetadata().getOrDefault(EXECUTION_THREAD_CPU_TIME_NS.getName(), "0")) + responseSerializationCpuTimeNs;
+    getMetadata().put(EXECUTION_THREAD_CPU_TIME_NS.getName(), String.valueOf(executionThreadCpuTimeNs));
+    // Copy all KV pair in _metadataV2 into _metadata
+    for (String key : _metadataV2.keySet()) {
+      Optional<MetadataKeys> opt = MetadataKeys.getByName(key);
+      if (!opt.isPresent()) {
+        continue;
+      }
+      _metadata.put(opt.get(), _metadataV2.get(key));
+    }
+    // Write metadata length and bytes.
+    byte[] metadataBytes = serializeMetadata();
+    dataOutputStream.writeInt(metadataBytes.length);
+    dataOutputStream.write(metadataBytes);
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  /**
+   * Serialize metadata section to bytes.
+   * Format of the bytes looks like:
+   * [numEntries, bytesOfKV2, bytesOfKV2, bytesOfKV3]
+   * For each KV pair:
+   * - if the value type is String, encode it as: [keyID, valueLength, Utf8EncodedValue].
+   * - if the value type is int, encode it as: [keyID, bigEndianRepresentationOfIntValue]
+   * - if the value type is long, encode it as: [keyID, bigEndianRepresentationOfLongValue]
+   */
+  private byte[] serializeMetadata()
+      throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
+
+    dataOutputStream.writeInt(_metadata.size());
+
+    for (Map.Entry<MetadataKeys, String> entry : _metadata.entrySet()) {
+      MetadataKeys key = entry.getKey();
+      String value = entry.getValue();
+      dataOutputStream.writeInt(key.ordinal());
+      if (MetadataKeys.isIntValueMetadataKey(key)) {
+        dataOutputStream.write(Ints.toByteArray(Integer.parseInt(value)));
+      } else if (MetadataKeys.isLongValueMetadataKey(key)) {
+        dataOutputStream.write(Longs.toByteArray(Long.parseLong(value)));
+      } else {
+        byte[] valueBytes = StringUtil.encodeUtf8(value);
+        dataOutputStream.writeInt(valueBytes.length);
+        dataOutputStream.write(valueBytes);
+      }
+    }
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  private Map<MetadataKeys, String> deserializeMetadata(byte[] bytes)
```

Comment:
(nit) Add a comment here highlighting something like  ---

Even though the wire format uses UTF-8 for string/bytes and big-endian for numeric values, the in-memory representation is STRING based for processing the metadata before serialization (by the server as it adds the statistics in metadata) and after deserialization (by the broker as it receives DataTable from each server and aggregates the values). This goes back to the point of how the V3 implementation has kept the consumers of `Map<String, String> getMetadata()` API in the code happy by internally converting to it.